### PR TITLE
Support for Swarm IDs

### PR DESCRIPTION
**Pull request by aluzzardi**
Swarm ID is an additional ID we have per container.

It's useful for a few reasons:
- We can know whether a container was created by Swarm or not
- User can address their containers by such ID
- It will remain constant in case of rebalancing

### CODE REVIEW

**Comment by aluzzardi** [comment]
Fixes #600

File: api/handlers.go [hunk 1]
```diff
@@ -251,6 +256,32 @@ func getContainerJSON(c *context, w http.ResponseWriter, r *http.Request) {
 		httpError(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
+	// Compatibility: Inject the Swarm ID label if not available.
+	// This **must** be done before inserting the `Node` field since it too has
+	// a `Label` field.
+	if !bytes.Contains(data, []byte("\"com.docker.swarm.id\":")) {
```
Comment (chanwit) [inline api/handlers.go]: probably
```
if !bytes.Contains(data, []byte(`"com.docker.swarm.id":`)) {
```
would be easier to read?

File: api/handlers.go [hunk 2]
```diff
@@ -251,6 +256,32 @@ func getContainerJSON(c *context, w http.ResponseWriter, r *http.Request) {
...
+		label := fmt.Sprintf("%q:%q", "com.docker.swarm.id", container.Config.SwarmID())
```
Comment (chanwit) [inline api/handlers.go]: Not sure if we need a constant for `"com.docker.swarm.id"`

File: cluster/config.go [hunk 3]
```diff
@@ -87,6 +87,17 @@ func (c *ContainerConfig) extractExprs(key string) []string {
 	return exprs
 }
 
+// SwarmID extracts the Swarm ID from the Config.
+// May return an empty string if not set.
+func (c *ContainerConfig) SwarmID() string {
+	return c.Labels[namespace+".id"]
```
Comment (chanwit) [inline cluster/config.go]: Oh right, I found the `namespace` const here. Is it possible to be reused in `api/handlers` as well?

File: cluster/config_test.go [hunk 4]
```diff
@@ -29,6 +29,18 @@ func TestBuildContainerConfig(t *testing.T) {
...
+	config := BuildContainerConfig(dockerclient.ContainerConfig{})
+	assert.Empty(t, config.SwarmID())
+	config.SetSwarmID("foo")
+	assert.Equal(t, config.SwarmID(), "foo")
+
+	// Retrieve an existing ID.
+	config = BuildContainerConfig(dockerclient.ContainerConfig{Labels: map[string]string{namespace + ".id": "test"}})
+	assert.Equal(t, config.SwarmID(), "test")
+}
+
```
Comment (chanwit) [inline cluster/config_test.go]: @aluzzardi  is it possible to add another test for the fake id? I saw you mentioned about the backward compatibility for 0.2.0.

File: cluster/config.go (hunk 3 above)
Comment (vieux) [inline cluster/config.go]: yes I need it too for another PR, I think we could do a `consts.go` in `/cluster` and export the variable there.

**Comment by aluzzardi** [comment]
Thanks for the feedback!

I decided to remove backward compatibility altogether.

It's not a huge deal, backward compatibility was meant to do stuff like rebalancing on containers created prior to 0.3, but since rebalancing will be an opt-in feature (there must be a flag passed during `docker run`), we won't rebalance older containers anyway so it's not necessary.

PTAL

**Comment by aluzzardi** [comment]
PTAL @vieux @abronan

**Comment by vieux** [comment]
LGTM

**Comment by abronan** [comment]
LGTM

**Pull request merged**